src/devices/bus/pc8801/pc8801_23.cpp

```diff
@@ -0,0 +1,89 @@
+// license:BSD-3-Clause
+// copyright-holders:Angelo Salese
+/**************************************************************************************************
+
+	NEC PC-8801-23 / -24 / -25 "Sound Board II"
+
+	TODO:
+	- Irq hangs if there's a PC-8801 OPN/OPNA declared in driver.
+      Master OPNA keeps sending acks in the correlated INT4 handler that hampers irq
+	  signals from here;
+	- Confirm there's no DB9 joyport on any variants of this;
+
+**************************************************************************************************/
+
+#include "emu.h"
+#include "pc8801_23.h"
+
+DEFINE_DEVICE_TYPE(PC8801_23, pc8801_23_device, "pc8801_23", "NEC PC-8801-23 \"Sound Board II\"")
+
+pc8801_23_device::pc8801_23_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock) 
+    : pc8801_exp_device(mconfig, PC8801_23, tag, owner, clock)
+	, m_opna(*this, "opna")
+{
+}
+
+void pc8801_23_device::io_map(address_map &map)
+{
+	map(0xa8, 0xa9).rw(m_opna, FUNC(ym2608_device::read), FUNC(ym2608_device::write));
+	map(0xaa, 0xaa).rw(FUNC(pc8801_23_device::irq_status_r), FUNC(pc8801_23_device::irq_mask_w));
+	map(0xac, 0xad).lrw8(
+		NAME([this] (offs_t offset) { return m_opna->read(offset | 2); }),
+		NAME([this] (offs_t offset, u8 data) { m_opna->write(offset | 2, data); })
+	);
+}
+
+void pc8801_23_device::opna_map(address_map &map)
+{
+	// TODO: confirm it really is ROMless
+	// TODO: confirm size
+	map(0x000000, 0x1fffff).ram();
+}
+
+void pc8801_23_device::device_add_mconfig(machine_config &config)
+{
+	constexpr XTAL sound_clock = XTAL(31'948'800) / 4;
+
+	YM2608(config, m_opna, sound_clock);
+	m_opna->set_addrmap(0, &pc8801_23_device::opna_map);
+	m_opna->irq_handler().set(FUNC(pc8801_23_device::int4_w));
+//	m_opna->port_a_read_callback().set(FUNC(pc8801_23_device::opn_porta_r));
+//	m_opna->port_b_read_callback().set_ioport("OPN_PB");
+	// TODO: per-channel mixing is unconfirmed
+	m_opna->add_route(0, "^^lspeaker", 0.25);
+	m_opna->add_route(0, "^^rspeaker", 0.25);
+	m_opna->add_route(1, "^^lspeaker", 0.75);
+	m_opna->add_route(2, "^^rspeaker", 0.75);
```

Comment:
This approach of making assumptions about something two levels away really isn’t supportable – the slot needs to implement `device_mixer_interface` with as many outputs as there are audio paths off the cards.

Reply:
Sure, can I have a documentation/howto about what `device_mixer_interface` is supposed to do?

Reply:
Basically, `device_mixer_interface` does automatic mixing/resampling of multiple audio streams.  You can see an example of converting a device to use it when @cam900 converted the NES SoC: in 3d0bb74c259f771c87c02bf8290183b05ff53b56 (#4846).  You can also see a simple example of it in some Zaccaria sound boards: https://github.com/mamedev/mame/blob/master/src/mame/audio/zaccaria.cpp

Basically:
* Make the slot derive from `device_mixer_interface`, pass the appropriate number of output channels to the `device_mixer_interface` constructor
* In the host system, add routes from the slot to the speakers or wherever it needs to go
* In the card, route sound to the slot using something like ` m_psg->add_route(<channel>, DEVICE_SELF_OWNER, <gain>, AUTO_ALLOC_INPUT, <output>);` where `<channel>` is the device’s output channel number, and `<output>` is the target mixer output it should end up on

The only nastiness is that you’ll probably need to `dynamic_cast` the card’s owner to ensure it’s actually the appropriate kind of slot before doing this to avoid breaking `-validate`, `-listroms`, `-listxml`, etc. as they depend on being able to instantiate devices in isolation.

Reply:
Actually, come to think of it, given that it doesn’t break `-valid` etc. now when it’s routing to speakers that don’t exist when the card is instantiated in isolation, you might not need to do a `dynamic_cast` to check at all.